crates/subspace-service/src/dsn/import_blocks/root_blocks.rs

```diff
@@ -0,0 +1,170 @@
+    pub async fn get_root_blocks(&self) -> Result<Vec<RootBlock>, Box<dyn Error>> {
+        trace!("Getting root blocks...");
+
+        let mut result = Vec::new();
+        let mut last_root_block = self.get_last_root_block().await?;
+        debug!(
+            "Getting root blocks starting from segment_index={}",
+            last_root_block.segment_index()
+        );
+
+        result.push(last_root_block);
+
+        while last_root_block.segment_index() > 0 {
+            let segment_indexes: Vec<_> = (0..last_root_block.segment_index())
+                .rev()
+                .take(ROOT_BLOCK_NUMBER_PER_REQUEST as usize)
+                .collect();
+
+            let mut root_blocks = self.get_root_blocks_batch(segment_indexes).await?;
+            root_blocks.sort_by(|rb1, rb2| rb1.segment_index().cmp(&rb2.segment_index()).reverse());
+
+            for root_block in root_blocks {
+                if root_block.hash() != last_root_block.prev_root_block_hash() {
+                    error!(
+                        hash=?root_block.hash(),
+                        prev_hash=?last_root_block.prev_root_block_hash(),
+                        "Root block hash doesn't match expected hash from the last block."
+                    );
+
+                    return Err(
+                        "Root block hash doesn't match expected hash from the last block.".into(),
+                    );
+                }
+
+                last_root_block = root_block;
+                result.push(root_block);
+            }
+        }
+
+        Ok(result)
+    }
+
+    async fn get_last_root_block(&self) -> Result<RootBlock, Box<dyn Error>> {
+        trace!("Getting last root block...");
+
+        // Get random peers. Some of them could be bootstrap nodes with no support for
+        // request-response protocol for records root.
+        let get_peers_result = self
+            .dsn_node
+            .get_closest_peers(PeerId::random().into())
+            .await;
+
+        match get_peers_result {
+            Ok(mut get_peers_stream) => {
+                while let Some(peer_id) = get_peers_stream.next().await {
+                    trace!(%peer_id, "get_closest_peers returned an item");
+
+                    let request_result = self
+                        .dsn_node
+                        .send_generic_request(
+                            peer_id,
+                            RootBlockRequest::LastRootBlocks {
+                                root_block_number: ROOT_BLOCK_NUMBER_PER_REQUEST,
+                            },
+                        )
+                        .await;
+
+                    match request_result {
+                        Ok(RootBlockResponse { root_blocks }) => {
+                            trace!(%peer_id, "Last root block request succeeded.");
+
+                            let last_root_block =
+                                root_blocks.iter().max_by_key(|rb| rb.segment_index());
```

Comment:
I want to have canonical order everywhere. If you accept it in reference implementation then every implementation going forward will have to implement it as well because there is a chance some implementation will send arbitrarily sorted garbage.

Why? What is the point of supporting non-canonical ordering? There is clearly a technical debt here because in all places where we make these requests you'll have to implement re-sorting of result again and again, but I do not understand what we're gaining in return.